src/ops/shapes.rs

```diff
@@ -0,0 +1,2528 @@
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let x = Array::from_float(2.3);
+/// let result = try_broadcast_to(&x, &[1, 1]);
+/// ```
+#[default_device]
+pub fn try_broadcast_to_device<'a>(
+    a: &'a Array,
+    shape: &'a [i32],
+    stream: StreamOrDevice,
+) -> Result<Array, BroadcastError<'a>> {
+    if !is_broadcastable(a.shape(), shape) {
+        return Err(BroadcastError {
+            src_shape: a.shape(),
+            dst_shape: shape,
+        });
+    }
+    unsafe { Ok(broadcast_to_device_unchecked(a, shape, stream)) }
+}
+
+/// Broadcast an array to the given shape. Panics if the shapes are not broadcastable.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `shape`: The shape to broadcast to.
+///
+/// # Panics
+///
+/// Panics if the shapes are not broadcastable. See [`try_broadcast_to`] for more information.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let x = Array::from_float(2.3);
+/// let y = broadcast_to(&x, &[1, 1]);
+/// ```
+#[default_device]
+pub fn broadcast_to_device<'a>(a: &'a Array, shape: &'a [i32], stream: StreamOrDevice) -> Array {
+    try_broadcast_to_device(a, shape, stream).unwrap()
+}
+
+fn concatenate_inner(arrays: &[impl AsRef<Array>], axis: i32, stream: StreamOrDevice) -> Array {
+    unsafe {
+        let c_arrays = new_mlx_vector_array(arrays);
+        let c_array = mlx_sys::mlx_concatenate(c_arrays, axis, stream.as_ptr());
+
+        let result = Array::from_ptr(c_array);
+        mlx_sys::mlx_free(c_arrays as *mut c_void);
+
+        result
+    }
+}
+
+/// Concatenate the arrays along the given axis.
+///
+/// # Params
+///
+/// - `arrays`: The arrays to concatenate.
+/// - `axis`: The axis to concatenate along.
+///
+/// # Safety
+///
+/// The function is unsafe because it does not check if the shapes are valid for concatenation.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let x = Array::from_iter(0..4, &[2, 2]);
+/// let y = Array::from_iter(4..8, &[2, 2]);
+/// let z = unsafe { concatenate_unchecked(&[x, y], 0) };
+/// ```
+#[default_device]
+pub unsafe fn concatenate_device_unchecked(
+    arrays: impl IntoIterator<Item = impl AsRef<Array>>,
+    axis: impl Into<Option<i32>>,
+    stream: StreamOrDevice,
+) -> Array {
+    let axis = axis.into().unwrap_or(0);
+    let arrays = arrays.into_iter().collect::<Vec<_>>();
+
+    concatenate_inner(&arrays, axis, stream)
+}
+
+/// Concatenate the arrays along the given axis. Returns an error if the shapes are invalid.
+///
+/// # Params
+///
+/// - `arrays`: The arrays to concatenate.
+/// - `axis`: The axis to concatenate along.
+///
+/// # Example
+///
+/// ```rust
+/// use mlx::{prelude::*, ops::*};
+///
+/// let x = Array::from_iter(0..4, &[2, 2]);
+/// let y = Array::from_iter(4..8, &[2, 2]);
+/// let result = try_concatenate(&[x, y], 0);
+/// ```
+#[default_device]
+pub fn try_concatenate_device(
```

Comment:
same as try_stack_device

Reply:
Addressed in 55ca17d931a3dff70b189ecb3086957543ca4d55